### PR TITLE
DPL-548: add received date to location report

### LINKED ISSUE #3693: DPL-nnn Location report include received date
**User story**
As a Sample Management user, I would like the plate location report to include the received date for each labware, so that I can assess in bulk which labware are old enough to be destroyed.

**Who are the primary contacts for this story**
Danni W, Katy

**Who is the nominated tester for UAT**
Danni W

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Plate location report (https://training.sequencescape.psd.sanger.ac.uk/location_reports) contains new column, 'Received Date'
- [ ] Received date is the date / time that the plate was scanned into Asset Audits, using the 'Receive plates' process, and the 'Reception fridge' instrument - see 'Additional context' for how to retrieve this information.

**References**
This story has a non-blocking relationship with other stories with the 'labware destruction' label, especially https://github.com/sanger/sequencescape/issues/3674, which also adds a column to the plate location report.

**Additional context**
Example of how to get the received date for a labware:
```
Labware.find(27412726).asset_audits.where(key: 'slf_receive_plates').last.created_at
```

### PR DESCRIPTION
**Pull request by StephenHulme**
Closes #3693 

#### Changes proposed in this pull request

- Add Received Date column to report along with update tests

#### Example

An extract showing the case where the recieved date is unknown:

```csv
"ScannedBarcode","HumanBarcode","Type","Created","ReceivedDate","Location","Service","RetentionInstructions","StudyName","StudyId","FacultySponsor"
"SQPD-9002","SQPD-9002","RVI Cherrypick","2023-10-09 13:45:22","Unknown","Not found - There is a problem with Labwhere","LabWhere","Unknown","A study","2","UAT Faculty Sponsor"
```

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: app/models/location_report.rb [hunk 1]
```diff
@@ -134,18 +146,29 @@ def plates_list
...
+      &.strftime('%Y-%m-%d %H:%M:%S') #TODO: add reception fridge check on message
+  end
+
   def generate_report_row(cur_plate, cur_study)
     row = generate_plate_cols_for_row(cur_plate)
     row + generate_study_cols_for_row(cur_study)
   end
 
-  def generate_plate_cols_for_row(cur_plate)
+  def generate_plate_cols_for_row(cur_plate) # rubocop:disable Metrics/AbcSize
```
[the remark this answers is not in the document]
Comment (sdjmchattie) [reply app/models/location_report.rb:164]: I suppose you could make a plate extension that, for a given plate, deals with the complexity of getting values that are sometimes Unknown as separate methods.  Then you could reduce the complexity in this method by just calling the methods one after another without the `||` handling.

File: app/models/location_report.rb [hunk 2]
```diff
@@ -134,18 +146,29 @@ def plates_list
...
 
+  def received_date(cur_plate)
+    cur_plate
+      &.asset_audits
+      &.where(key: 'slf_receive_plates')
+      &.where('message LIKE ?', '%Reception fridge%')
+      &.first
+      &.created_at
+      &.strftime('%Y-%m-%d %H:%M:%S')
+  end
```
Comment (KatyTaylor) [inline app/models/location_report.rb:157]: Could be worth a question to Danni - if it has been 'received' multiple times, whether to use the first or the last event. Although hopefully it's an edge case.

**Review by KatyTaylor — APPROVED** [review]
Looks good to me, just one question around first or last asset audit. Get it deployed in UAT (if poss!) and ask Danni to test it? 

Did you manage to generate it yourself manually, with some test data? 

Thanks

File: app/models/location_report.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdjmchattie) [reply app/models/location_report.rb]: Nice!  Anything that avoids turning off the rubocop rule is a win.  I know the change you've made isn't super easily solved, but once it's off, people can come in and make a hideous change and not realise how bad it is.

**Review by sdjmchattie — APPROVED** [review]
